// anycontrol/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
/* eslint-disable no-undef */
/* eslint-disable no-console */
/* eslint-disable no-lonely-if */

var AnyControl = function AnyControl() {
  var _this = this;

  var SpeechRecognition = window.webkitSpeechRecognition || window.SpeechRecognition;
  this.recognition = new SpeechRecognition();
  this.finalTranscript = '';
  this.commands = {};
  this.DEBUG = false;
  this.recognizing = false;
  this.recognition.onresult = function (event) {
    if (typeof event.results === 'undefined') {
      if (_this.DEBUG) {
        console.info('undefined result');
      }

      _this.recognition.stop();
      return;
    }

    for (var i = event.resultIndex; i < event.results.length; i += 1) {
      if (event.results[i].isFinal) {
        _this.finalTranscript += event.results[i][0].transcript;
      }
    }

    if (_this.finalTranscript !== '') {
      if (_this.DEBUG) {
        console.info('received command:', _this.finalTranscript);
      }

      _this.finalTranscript = _this.finalTranscript.toLowerCase().trim();
      var commands = _this.commands;
      Object.keys(commands).forEach(function (command) {
        if (_this.finalTranscript.indexOf(command) > -1) {
          if (_this.finalTranscript[command.length] === undefined) {
            if (_this.DEBUG) {
              console.info('calling command', command);
            }
            _this.commands[command]();
          } else if (_this.finalTranscript[command.length] === ' ') {
            var param = _this.finalTranscript.substring(command.length, _this.finalTranscript.length).trim();
            _this.commands[command](param);

            if (_this.DEBUG) {
              console.info('calling command', command, 'with param:', param);
            }
          }
        }
      });

      _this.finalTranscript = '';
    } else {
      if (_this.DEBUG) {
        console.info('received empty command');
      }
    }
  };

  this.recognition.onerror = function (event) {
    if (_this.DEBUG) {
      console.info('error occured', event);
    }
  };

  this.recognition.onstart = function (event) {
    if (_this.DEBUG) {
      console.info('start', event);
    }

    _this.recognizing = true;
  };

  this.recognition.onend = function (event) {
    if (_this.DEBUG) {
      console.info('end', event);
    }

    _this.recognizing = false;

    if (_this.recognition.continuous) {
      if (_this.DEBUG) {
        console.info('restarting', _this.recognition.continuous);
      }

      _this.recognition.start();
    }
  };

  return this;
};

AnyControl.prototype.isSupported = function isSupported() {
  return !!this.recognition;
};

AnyControl.prototype.addCommand = function addCommand(command, callback) {
  if (this.DEBUG) {
    console.info('added command:', command);
  }

  this.commands[command.toLowerCase()] = callback;
};

AnyControl.prototype.removeCommand = function removeCommand(command) {
  if (this.DEBUG) {
    console.info('removed command:', command);
  }

  if (this.commands[command]) {
    delete this.commands[command];
    return true;
  }

  return false;
};

AnyControl.prototype.start = function start() {
  if (this.DEBUG) {
    console.info('started listening');
  }

  this.recognition.continuous = true;
  this.recognition.start();
};

AnyControl.prototype.stop = function stop() {
  if (this.DEBUG) {
    console.info('stopped listening');
  }

  this.recognition.continuous = false;
  this.recognition.stop();
};

AnyControl.prototype.getCommand = function getCommand() {
  var _this2 = this;

  var timeout = 1;

  if (this.isRecognizing()) {
    this.stop();

    timeout = 1000;
  }

  setTimeout(function () {
    if (_this2.DEBUG) {
      console.info('listening for single command');
    }

    _this2.recognition.continuous = false;
    _this2.recognition.start();
  }, timeout);
};

AnyControl.prototype.debug = function debug(mode) {
  this.DEBUG = !!mode;
};

AnyControl.prototype.isRecognizing = function isRecognizing() {
  return this.recognizing;
};

var main = AnyControl;
exports.default = main;

module.exports = main; // for CommonJS compatibility
